### PR TITLE
authn: Resave an on-the-fly updated user object back to the stored session

### PR DESCRIPTION
**Pull request by tsibley**
Rather than perpetually updating it on-the-fly with every request.

Resolves a downside of "authn: Upgrade existing sessions which pre-date
addition of authzRoles" (272ed09) and was simpler to implement than I
thought!

### Related issue(s)  

Related to #479 and #452.

### Testing
I tested locally (which uses a file-based session store) by first acquiring an "old" session (i.e. logging in on an earlier version of the codebase), which contained a user object like:

```console
$ jq '.passport.user | fromjson' sessions/lMG7Ra3e79V5LuefCy2ObeMzucT4q5u6.json  
{
  "username": "trs",
  "groups": [
    "test/owners",
    "blab-private",
    "seattleflu",
    "trs-test/owners",
    "test-private/editors",
    "blab",
    "inrb-drc"
  ]
}
```

then I restarted my local server with the code in this PR and made another request, after which my session contained the expected updates: 

```console
$ jq '.passport.user | fromjson' sessions/lMG7Ra3e79V5LuefCy2ObeMzucT4q5u6.json  
{
  "username": "trs",
  "groups": [
    "test",
    "blab-private",
    "seattleflu",
    "trs-test",
    "test-private",
    "blab",
    "inrb-drc"
  ],
  "authzRoles": [
    "test/owners",
    "blab-private/viewers",
    "seattleflu/viewers",
    "trs-test/owners",
    "test-private/editors",
    "blab/viewers",
    "inrb-drc/viewers"
  ]
}
```

### CODE REVIEW

File: src/authn/index.js [hunk 1]
```diff
@@ -218,6 +225,19 @@ function setup(app) {
   // be null.
   app.use(passport.session());
 
+  // If the user object was modified in deserializeUser during restore from the
+  // session, then persist the modifications back into the session by updating
+  // the user object in the session via req.login().  The session will detect
+  // its been changed and save itself to the store at the end of the request.
+  app.use((req, res, next) => {
+    if (req.user?.[RESAVE_TO_SESSION]) {
```
Comment (joverlee521) [inline src/authn/index.js:233]: Wow, didn't know [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) was a thing! 🤓

**Review by joverlee521 — APPROVED** [review]
Very cool!

File: src/authn/index.js (hunk 1 above)
Comment (tsibley) [reply src/authn/index.js:233]: We couldn't use it here until the recent version bump from Node 13 → 14, so it's a nice perk of that! In the same family is also `??`, the "[nullish coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator)".

**Comment by tsibley** [comment]
Deployed and working:

```
Mar 08 10:50:03 nextstrain-server app/web.3 [verbose]	Upgrading groups and authzRoles of user object for trs
Mar 08 10:50:03 nextstrain-server app/web.3 [verbose]	Resaving user object for trs back to the session
```

**Pull request merged**